app/src/main/java/com/example/candyshop/api/CandyItemsRepository.kt

```diff
@@ -0,0 +1,82 @@
+package com.example.candyshop.api
+
+import com.example.candyshop.data.db.CandyDatabase
+import com.example.candyshop.data.model.Candy
+import com.example.candyshop.data.Resource
+import com.example.candyshop.data.toCandy
+import com.example.candyshop.data.toCandyEntity
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.flow
+import retrofit2.HttpException
+import java.io.IOException
+import javax.inject.Inject
+
+interface CandyItemsRepository {
+    suspend fun getCandyItems(
+        forceFetchFromRemote: Boolean
+    ): Flow<Resource<List<Candy>>>
+
+    suspend fun getCandy(id: Int): Flow<Resource<Candy>>
+}
+
+class CandyItemsRepositoryImplementation @Inject constructor(
+    private val candyApiService: CandyApiService,
+    private val candyDatabase: CandyDatabase
+) : CandyItemsRepository {
+    override suspend fun getCandyItems(forceFetchFromRemote: Boolean): Flow<Resource<List<Candy>>> {
+        return flow {
+            emit(Resource.Loading(true))
+            val localCandyList = candyDatabase.candyDao.getAllCandy()
+            val shouldLoadLocalCandy = localCandyList.isNotEmpty() && !forceFetchFromRemote
+
+            if (shouldLoadLocalCandy) {
+                emit(Resource.Success(data = localCandyList.map { candyEntity ->
+                    candyEntity.toCandy()
+                }))
+                emit(Resource.Loading(false))
+                return@flow
+            }
+
+            val candyListFromApi = try {
+                candyApiService.getDesserts()
+            } catch (e: IOException) {
+                e.printStackTrace()
+                emit(Resource.Error(message = "Error loading desserts. IOException"))
+                return@flow
+            } catch (e: HttpException) {
+                e.printStackTrace()
+                emit(Resource.Error(message = "Error loading desserts. HTTPException"))
+                return@flow
+            } catch (e: Exception) {
+                e.printStackTrace()
+                emit(Resource.Error(message = "Error loading desserts. GeneralException"))
+                return@flow
+            }
+
+            val candyEntities = candyListFromApi.meals.let {
+                it.map { candyDto ->
+                    candyDto.toCandyEntity()
+                }
+            }
+
+            candyDatabase.candyDao.upsertCandyList(candyEntities)
```

Comment:
also operation with db can throw exception so would be good to catch also this one